Added support for DFU of wi-fi firmware patch (#14075)
nrfconnect/sdk-nrf: samples/matter/lock/README.rst

See `Matter door lock build types`_, `Selecting a build type`_, and :ref:`matter_lock_sample_switching_thread_wifi` for more information about how to configure and test this feature with this sample.

.. matter_door_lock_sample_nrf70_firmware_patch_start

A portion of the application code related to the nRF70 Wi-Fi firmware can be flashed into external memory to free up space in the on-chip memory.
This option is available only when building for the nRF5340 DK with the nRF7002 EK shield attached.
To prepare an application to use this feature, you need to create additional MCUBoot partitions.

To learn how to configure MCUBoot partitions, see the :ref:`nrf70_fw_patch_update_adding_partitions` guide.

> To enable this feature, set the :kconfig:option:`CONFIG_NRF_WIFI_FW_PATCH_DFU` Kconfig option for both the Application and the MCUBoot child image.

Reviewer: You no longer need that in the MCUBoot child image, right?

Author: Right, changed.

You can do this by completing one of the following actions:

* Add ``CONFIG_NRF_WIFI_FW_PATCH_DFU=y`` to the prj.conf file for the Application, and ``CONFIG_UPDATEABLE_IMAGE_NUMBER=3`` to the ``child_image/mcuboot/prj.conf`` file for the MCUBoot child image.

* Build the sample with the additional arguments ``-DCONFIG_NRF_WIFI_PATCHES_EXT_FLASH_STORE=y`` for the Application, and ``-Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3`` for the MCUBoot child image.

.. matter_door_lock_sample_nrf70_firmware_patch_end

For example:

   .. code-block:: console

>       west build -d nrf5340dk_nrf5340_cpuapp -d -- -DSHIELD=nrf7002ek -Dmultiprotocol_rpmsg_SHIELD=nrf7002ek_coex -DCONF_FILE=prj_thread_wifi_switched.conf -DCONFIG_NRF_WIFI_PATCHES_EXT_FLASH_STORE=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3

Reviewer 2: Do not add this information in the sample docs. This should be explained in the user guide in the configuration section (`nrf70_fw_patch_update.rst`) and just linked to here.

> .. matter_door_lock_sample_nrf70_firmware_patch_start
> 
> A portion of the application code related to the nRF70 Wi-Fi firmware can be flashed into external memory to free up space in the on-chip memory.
> This option is available only when building for the nRF5340 DK with the nRF7002 EK shield attached.
> To prepare an application to use this feature, you need to create additional MCUboot partitions.
> 
> To learn how to configure MCUboot partitions, see the :ref:`nrf70_fw_patch_update_adding_partitions` guide.
> 
> To enable this feature, set the :kconfig:option:`CONFIG_NRF_WIFI_FW_PATCH_DFU` Kconfig option to ``y`` for the Application and set the :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` kconfig option to ``3`` for the MCUBoot child image.
> You can do this by completing one of the following actions:
> 
> * Add ``CONFIG_NRF_WIFI_FW_PATCH_DFU=y`` to the :file:`prj.conf` configuration file for the application, and ``CONFIG_UPDATEABLE_IMAGE_NUMBER=3`` to the :file:`child_image/mcuboot/prj.conf` configuration file for the MCUboot child image.
> 
> * Build the sample with the additional arguments ``-DCONFIG_NRF_WIFI_PATCHES_EXT_FLASH_STORE=y`` for the application, and ``-Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3`` for the MCUboot child image.
> 
> .. matter_door_lock_sample_nrf70_firmware_patch_end
> 
> For example:
> 
>    .. code-block:: console
> 
>       west build -d nrf5340dk_nrf5340_cpuapp -d -- -DSHIELD=nrf7002ek -Dmultiprotocol_rpmsg_SHIELD=nrf7002ek_coex -DCONF_FILE=prj_thread_wifi_switched.conf -DCONFIG_NRF_WIFI_PATCHES_EXT_FLASH_STORE=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3
> 

Reviewer 2: ```suggestion
Wi-Fi firmware on external memory
---------------------------------

.. matter_door_lock_sample_nrf70_firmware_patch_start

You can program a portion of the application code related to the nRF70 Series' Wi-Fi firmware onto an external memory to free up space in the on-chip memory.
This option is available only when building for the nRF5340 DK with the nRF7002 EK shield attached.

To prepare an application to use this feature, you need to create additional MCUboot partitions.
To learn how to configure MCUboot partitions, see the :ref:`nrf70_fw_patch_update_adding_partitions` guide.

To enable this feature for Matter, set the :kconfig:option:`CONFIG_NRF_WIFI_FW_PATCH_DFU` Kconfig option to ``y`` for the application (in the application :file:`prj.conf`) and the :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` Kconfig option to ``3`` for the MCUBoot child image (in its own :file:`prj.conf`).

.. matter_door_lock_sample_nrf70_firmware_patch_end

For example:

   .. code-block:: console

      west build -d nrf5340dk_nrf5340_cpuapp -d -- -DSHIELD=nrf7002ek -Dmultiprotocol_rpmsg_SHIELD=nrf7002ek_coex -DCONF_FILE=prj_thread_wifi_switched.conf -DCONFIG_NRF_WIFI_PATCHES_EXT_FLASH_STORE=y -Dmcuboot_CONFIG_UPDATEABLE_IMAGE_NUMBER=3

```
